ttnn/cpp/ttnn/operations/ccl/common/kernels/ccl_send_writer.cpp

```diff
@@ -0,0 +1,274 @@
+constexpr uint32_t cb_id = get_compile_time_arg_val(3);
+
+#ifdef SHARDED_MEM_LAYOUT
+static constexpr bool is_sharded_mode = true;
+static constexpr uint32_t input_tensor_shard_grid_height = get_compile_time_arg_val(4);
+static constexpr uint32_t input_tensor_shard_grid_width = get_compile_time_arg_val(5);
+static constexpr uint32_t input_tensor_shard_grid_start_y_logical = get_compile_time_arg_val(6);
+static constexpr uint32_t input_tensor_shard_grid_start_x_logical = get_compile_time_arg_val(7);
+static constexpr uint32_t input_tensor_shard_pages_per_shard_y = get_compile_time_arg_val(8);
+static constexpr uint32_t input_tensor_shard_pages_per_shard_x = get_compile_time_arg_val(9);
+static constexpr bool input_tensor_shard_grid_transposed = get_compile_time_arg_val(10) != 0;
+#else
+static constexpr bool is_sharded_mode = false;
+static constexpr uint32_t input_tensor_shard_grid_height = 0;
+static constexpr uint32_t input_tensor_shard_grid_width = 0;
+static constexpr uint32_t input_tensor_shard_grid_start_y_logical = 0;
+static constexpr uint32_t input_tensor_shard_grid_start_x_logical = 0;
+static constexpr uint32_t input_tensor_shard_pages_per_shard_y = 0;
+static constexpr uint32_t input_tensor_shard_pages_per_shard_x = 0;
+static constexpr bool input_tensor_shard_grid_transposed = false;
+#endif
+
+template <
+    tt::tt_metal::TensorMemoryLayout tensor_layout,
+    tt::tt_metal::BufferType buffer_type,
+    tt::tt_metal::Layout page_layout>
+auto build_source_address_generator(
+    std::size_t& arg_idx, address_t tensor_address, std::size_t page_size, uint32_t cb_id_in0) ->
+    typename source_tensor_addrgen<tensor_layout, buffer_type, page_layout>::type {
+    constexpr bool is_sharded = is_sharded_tensor_layout(tensor_layout);
+    constexpr bool is_interleaved = tensor_layout == tt::tt_metal::TensorMemoryLayout::INTERLEAVED;
+    constexpr bool is_tile_page_layout = page_layout == tt::tt_metal::Layout::TILE;
+    constexpr bool is_row_major_layout = page_layout == tt::tt_metal::Layout::ROW_MAJOR;
+    static_assert(
+        is_sharded || is_interleaved,
+        "Only sharded and interleaved tensor layouts are supported but the unified address generator. A tensor layout "
+        "not matching TensorMemoryLayout::WIDTH_SHARDED, TensorMemoryLayout::HEIGHT_SHARDED, "
+        "TensorMemoryLayout::BLOCK_SHARDED, or TensorMemoryLayout::INTERLEAVED was specified.");
+
+    using addrgen_type = typename source_tensor_addrgen<tensor_layout, buffer_type, page_layout>::type;
+
+    if constexpr (tensor_layout == tt::tt_metal::TensorMemoryLayout::INTERLEAVED) {
+        if constexpr (is_row_major_layout) {
+            return addrgen_type{.bank_base_address = tensor_address, .page_size = page_size};
+        } else {
+            return addrgen_type{
+                .bank_base_address = tensor_address, .page_size = page_size, .data_format = get_dataformat(cb_id_in0)};
+        }
+    } else if constexpr (
```

Comment:
Same comment as before regarding that we may be able to clean up all those default constant expression values if we use #if and #else notation